metal/drivers/sifive_ccache0.h

```diff
@@ -1,22 +1,123 @@
-/* Copyright 2019 SiFive, Inc */
+/* Copyright 2020 SiFive, Inc */
 /* SPDX-License-Identifier: Apache-2.0 */
 
 #ifndef METAL__DRIVERS__SIFIVE_CCACHE0_H
 #define METAL__DRIVERS__SIFIVE_CCACHE0_H
 
-#include <metal/cache.h>
-#include <metal/compiler.h>
+/*!
+ * @file sifive_ccache0.h
+ *
+ * @brief API for configuring the SiFive L2 cache controller
+ */
 
-struct __metal_driver_vtable_sifive_ccache0 {
-    struct __metal_cache_vtable cache;
-};
+#include <stdint.h>
 
-struct __metal_driver_sifive_ccache0;
+/*! @brief Cache configuration data */
+typedef struct {
+    uint32_t num_bank;
+    uint32_t num_ways;
+    uint32_t num_sets;
+    uint32_t block_size;
+} sifive_ccache0_config;
 
-__METAL_DECLARE_VTABLE(__metal_driver_vtable_sifive_ccache0)
+/*! @brief Set of values for ECC error type*/
+typedef enum {
+    SIFIVE_CCACHE0_DATA = 0,
+    SIFIVE_CCACHE0_DIR = 1,
+} sifive_ccache0_ecc_errtype_t;
 
-struct __metal_driver_sifive_ccache0 {
-    struct metal_cache cache;
-};
+/*! @brief Initialize cache controller with specified ways.
+ * @param ways Number of ways to be enabled.
+ * @return 0 If no error.*/
+int sifive_ccache0_init(uint32_t ways);
+
+/*! @brief Get cache configuration data.
+ * @param None.
+ * @return Cache controller configuration data.*/
+sifive_ccache0_config *sifive_ccache0_get_config(void);
```

Comment:
Instead of returning the pointer to a struct with these values (where does that even get allocated?), maybe just have four functions, one for each value? Especially since there doesn't appear to be a "set config" function, I don't see the need for the struct.

Reply:
Function now takes a struct param from the user. This API returns read only static cache configuration data, struct is just more condensed way of returning this data to the user. Adding few more functions might unnecessarily add to already long list of driver functions.